File: src/main/nashorn/net/x3djsonld/data/PixelTextureComponentExamples.js
load('X3Dautoclass.js');

// Javadoc annotations follow, see below for source.
/**
 * <p> This example shows the five PixelTexture components, with 0 to 4 components each, shown in Table 5-18. </p>
 <p> Related links: <a href="../../../Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.java">PixelTextureComponentExamples.java</a> source, <a href="../../../Chapter05AppearanceMaterialTextures/PixelTextureComponentExamplesIndex.html" target="_top">PixelTextureComponentExamples catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.x3d">PixelTextureComponentExamples.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> This example shows the five PixelTexture components, with 0 to 4 components each, shown in Table 5-18. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Leonard Daly and Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 25 August 2008 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 7 January 2014 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="http://X3dGraphics.com" target="_blank">http://X3dGraphics.com</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> X3D for Web Authors, Table 5.18 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="http://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">http://www.web3d.org/x3d/content/examples/X3dResources.html</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> rights </i> </td>
			<td> Copyright (c) 2006, Daly Realism and Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> subject </i> </td>
			<td> X3D book, X3D graphics, X3D-Edit, <a href="http://www.x3dGraphics.com" target="_blank">http://www.x3dGraphics.com</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.x3d" target="_blank">http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="../../../Chapter05AppearanceMaterialTextures/../license.html">../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToES5.xslt" target="_blank">X3dToES5.xslt</a>
		stylesheet to create ES5 source code from an <code>.x3d</code> scene.
	</p>

	* @author Leonard Daly and Don Brutzman
 */

function PixelTextureComponentExamples
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
PixelTextureComponentExamples.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {

	/** Large attribute array: PixelTexture image field, scene-graph level=6, element #68, 67 total numbers */
	this.PixelTexture_6_68_image = new SFImage(Java.to([8,8,1,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc], Java.type("int[]")));
  this.x3dModel = new X3D().setProfile("Immersive").setVersion("3.3")
  .setHead(new head()
    .addMeta(new meta().setName("title").setContent("PixelTextureComponentExamples.x3d"))
    .addMeta(new meta().setName("description").setContent("This example shows the five PixelTexture components, with 0 to 4 components each, shown in Table 5-18."))
    .addMeta(new meta().setName("creator").setContent("Leonard Daly and Don Brutzman"))
    .addMeta(new meta().setName("created").setContent("25 August 2008"))
    .addMeta(new meta().setName("modified").setContent("7 January 2014"))
    .addMeta(new meta().setName("reference").setContent("http://X3dGraphics.com"))
    .addMeta(new meta().setName("reference").setContent("X3D for Web Authors, Table 5.18"))
    .addMeta(new meta().setName("reference").setContent("http://www.web3d.org/x3d/content/examples/X3dResources.html"))
    .addMeta(new meta().setName("rights").setContent("Copyright (c) 2006, Daly Realism and Don Brutzman"))
    .addMeta(new meta().setName("subject").setContent("X3D book, X3D graphics, X3D-Edit, http://www.x3dGraphics.com"))
    .addMeta(new meta().setName("identifier").setContent("http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.x3d"))
    .addMeta(new meta().setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta(new meta().setName("license").setContent("../license.html")))
  .setScene(new Scene()
    .addChild(new Background().setSkyColor(new MFColor(Java.to([0.1,0.1,0.4], Java.type("float[]")))))
    .addChild(new Viewpoint().setDescription("Table 5.18 SFImage component examples").setPosition(0.0,0.0,14.0))
    .addChild(new Transform().setTranslation(-6.0,0.0,0.0)
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .setTexture(new PixelTexture("ZeroComponents")))
        .setGeometry(new Box()))
      .addChild(new Transform().setTranslation(0.0,-2.0,0.0)
        .addChild(new Shape()
          .setGeometry(new Text().setString(new MFString("\"0\""))
            .setFontStyle(new FontStyle("CenterJustify").setJustify(new MFString("\"MIDDLE\" \"MIDDLE\""))))
          .setAppearance(new Appearance("TextMaterial")
            .setMaterial(new Material().setDiffuseColor(1.0,1.0,1.0))))))
    .addChild(new Transform().setTranslation(-3.0,0.0,0.0)
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .setTexture(new PixelTexture("OneComponent").setImage(Java.to([1,2,1,0xFF,0x00], Java.type("int[]")))))
        .setGeometry(new Box()))
      .addChild(new Transform().setTranslation(0.0,-2.0,0.0)
        .addChild(new Shape()
          .setGeometry(new Text().setString(new MFString("\"1\""))
            .setFontStyle(new FontStyle().setUSE("CenterJustify")))
          .setAppearance(new Appearance().setUSE("TextMaterial")))))
    .addChild(new Transform()
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .setTexture(new PixelTexture("TwoComponents").setImage(Java.to([2,1,2,0xCCFF,0x2277], Java.type("int[]")))))
        .setGeometry(new Box()))
      .addChild(new Transform().setTranslation(0.0,-2.0,0.0)
        .addChild(new Shape()
          .setGeometry(new Text().setString(new MFString("\"2\""))
            .setFontStyle(new FontStyle().setUSE("CenterJustify")))
          .setAppearance(new Appearance().setUSE("TextMaterial")))))
    .addChild(new Transform().setTranslation(3.0,0.0,0.0)
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .addComments(" note 0x000000 = 0 ")
          .setTexture(new PixelTexture("ThreeComponents").setImage(Java.to([2,4,3,0xFF0000,0xFF00,0x000000,0,0,0,0xFFFFFF,0xFFFF00], Java.type("int[]")))))
        .setGeometry(new Box()))
      .addChild(new Transform().setTranslation(0.0,-2.0,0.0)
        .addChild(new Shape()
          .setGeometry(new Text().setString(new MFString("\"3\""))
            .setFontStyle(new FontStyle().setUSE("CenterJustify")))
          .setAppearance(new Appearance().setUSE("TextMaterial")))))
    .addChild(new Transform().setTranslation(6.0,0.0,0.0)
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .addComments(" Erroneous value in book: 1 0 0 255, 0 1 0 255, 0 0 1 255, 1 0 0 127, 0 1 0 127, 0 0 1 127 ")
          .setTexture(new PixelTexture("FourComponents").setImage(Java.to([3,2,4,0xFF0000FF,0x00FF00FF,0x0000FFFF,0xFF00007F,0x00FF007F,0x0000FF7F], Java.type("int[]")))))
        .setGeometry(new Box()))
      .addChild(new Transform().setTranslation(0.0,-2.0,0.0)
        .addChild(new Shape()
          .setGeometry(new Text().setString(new MFString("\"4\""))
            .setFontStyle(new FontStyle().setUSE("CenterJustify")))
          .setAppearance(new Appearance().setUSE("TextMaterial")))))
    .addComments(" Background from PixelTextureBW.x3d ")
    .addChild(new Transform().setTranslation(0.0,6.0,-2.0)
      .addChild(new Shape()
        .setAppearance(new Appearance()
          .setTexture(new PixelTexture().setImage(this.PixelTexture_6_68_image)))
        .setGeometry(new Box().setSize(16.0,16.0,.1)))));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return PixelTextureComponentExamples model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var       metaResult = "";
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		// first list informational meta elements of interest
		var metaList = this.getX3dModel().getHead().getMetaList();
		for (var m in metaList) {
			meta = metaList[m];
			if (meta.getName().equals(meta.NAME_ERROR) ||
				meta.getName().equals(meta.NAME_WARNING) ||
				meta.getName().equals(meta.NAME_HINT) ||
				meta.getName().equals(meta.NAME_INFO) ||
				meta.getName().equals(meta.NAME_TODO))
			{
				metaResult += meta.toStringX3D();
			}
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (metaResult === "" && exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = metaResult;
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new PixelTextureComponentExamples();
		print ("PixelTextureComponentExamples execution self-validation test results: " + testObject.validateSelf());
	}
}
new PixelTextureComponentExamples().main();

File: src/main/python/net/x3djsonld/data/PixelTextureComponentExamples.py
####################################################################################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix,
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.
#
####################################################################################################

from x3d import *

newModel=X3D(profile='Immersive',version='3.3',
  head=head(
    children=[
    meta(content='PixelTextureComponentExamples.x3d',name='title'),
    meta(content='This example shows the five PixelTexture components, with 0 to 4 components each, shown in Table 5-18.',name='description'),
    meta(content='Leonard Daly and Don Brutzman',name='creator'),
    meta(content='25 August 2008',name='created'),
    meta(content='7 January 2014',name='modified'),
    meta(content='http://X3dGraphics.com',name='reference'),
    meta(content='X3D for Web Authors, Table 5.18',name='reference'),
    meta(content='https://www.web3d.org/x3d/content/examples/X3dResources.html',name='reference'),
    meta(content='Copyright (c) 2006, Daly Realism and Don Brutzman',name='rights'),
    meta(content='X3D book, X3D graphics, X3D-Edit, http://www.x3dGraphics.com',name='subject'),
    meta(content='http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter05AppearanceMaterialTextures/PixelTextureComponentExamples.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../license.html',name='license')]),
  Scene=Scene(
    children=[
    Background(skyColor=[(0.1,0.1,0.4)]),
    Viewpoint(description='Table 5.18 SFImage component examples',position=(0,0,14)),
    Transform(translation=(-6,0,0),
      children=[
      Shape(
        appearance=Appearance(
          texture=PixelTexture(DEF='ZeroComponents')),
        geometry=Box(),),
      Transform(translation=(0,-2,0),
        children=[
        Shape(
          geometry=Text(string=["0"],
            fontStyle=FontStyle(DEF='CenterJustify',justify=["MIDDLE","MIDDLE"])),
          appearance=Appearance(DEF='TextMaterial',
            material=Material(diffuseColor=(1,1,1))))])]),
    Transform(translation=(-3,0,0),
      children=[
      Shape(
        appearance=Appearance(
          texture=PixelTexture(DEF='OneComponent',image=[1,2,1,0xFF,0x00])),
        geometry=Box(),),
      Transform(translation=(0,-2,0),
        children=[
        Shape(
          geometry=Text(string=["1"],
            fontStyle=FontStyle(USE='CenterJustify')),
          appearance=Appearance(USE='TextMaterial'))])]),
    Transform(
      children=[
      Shape(
        appearance=Appearance(
          texture=PixelTexture(DEF='TwoComponents',image=[2,1,2,0xCCFF,0x2277])),
        geometry=Box(),),
      Transform(translation=(0,-2,0),
        children=[
        Shape(
          geometry=Text(string=["2"],
            fontStyle=FontStyle(USE='CenterJustify')),
          appearance=Appearance(USE='TextMaterial'))])]),
    Transform(translation=(3,0,0),
      children=[
      Shape(
        appearance=Appearance(
          #  note 0x000000 = 0 
          texture=PixelTexture(DEF='ThreeComponents',image=[2,4,3,0xFF0000,0xFF00,0x000000,0,0,0,0xFFFFFF,0xFFFF00])),
        geometry=Box(),),
      Transform(translation=(0,-2,0),
        children=[
        Shape(
          geometry=Text(string=["3"],
            fontStyle=FontStyle(USE='CenterJustify')),
          appearance=Appearance(USE='TextMaterial'))])]),
    Transform(translation=(6,0,0),
      children=[
      Shape(
        appearance=Appearance(
          #  Erroneous value in book: 1 0 0 255, 0 1 0 255, 0 0 1 255, 1 0 0 127, 0 1 0 127, 0 0 1 127 
          texture=PixelTexture(DEF='FourComponents',image=[3,2,4,0xFF0000FF,0x00FF00FF,0x0000FFFF,0xFF00007F,0x00FF007F,0x0000FF7F])),
        geometry=Box(),),
      Transform(translation=(0,-2,0),
        children=[
        Shape(
          geometry=Text(string=["4"],
            fontStyle=FontStyle(USE='CenterJustify')),
          appearance=Appearance(USE='TextMaterial'))])]),
    #  Background from PixelTextureBW.x3d 
    Transform(translation=(0,6,-2),
      children=[
      Shape(
        appearance=Appearance(
          texture=PixelTexture(image=[8,8,1,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0x00,0xcc,0x00,0xcc,0x00,0xcc,0x00,0xcc])),
        geometry=Box(size=(16,16,.1)))])])
) # X3D model complete

####################################################################################################
# Self-test diagnostics
####################################################################################################

print('Self-test diagnostics for PixelTextureComponentExamples.py:')
if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel)) # display meta info, hint, warning, error, TODO values in this model
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()
# print(newModelXML) # diagnostic

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful", flush=True)
except Exception as err: # usually BaseException
    # https://stackoverflow.com/questions/18176602/how-to-get-the-name-of-an-exception-that-was-caught-in-python
    print("*** Python-to-VRML export of VRML output failed:", type(err).__name__, err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (under development)")
except Exception as err: # usually SyntaxError
    print("*** Python-to-JSON export of JSON output failed:", type(err).__name__, err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python PixelTextureComponentExamples.py load and self-test diagnostics complete.")
